### DISCUSSION TITLE
Clarification regarding "hack"

### DISCUSSION THREAD

**Discussion by StefanBrand** [s-weigand/flake8-nb#244] [Q&A] [answered]
Hello, this project looks really interesting, but to me personally the use of "hack" in the README puts me off.

> Basically this is a hack on the flake8's Application class

I propose to have this rewritten to 

> Basically this subclasses flake8's Application class

If you agree, I'm happy to create a PR. :) 

Edit: I see there are a lot of references to "hacks" in the source code. Still, the badges are quite convincing, so maybe "hack" is an understatement.

![image](https://user-images.githubusercontent.com/8948865/112120106-40138400-8bbe-11eb-8079-376507189865.png)

**Comment by s-weigand** [comment, marked as answer]
The reason why I call it a `hack` is that the `Application` class of `flake8` isn't part of the official API and it isn't supposed to be subclassed ( [see this comment from the creator of flake8](https://gitlab.com/pycqa/flake8/-/merge_requests/359#note_226407899) )

The problem is that notebooks need to be a kind of Schrödiger-state, ignored and used at the same time.
`flake8` can't be used directly on a notebook, since it internally is just a big JSON file, thus this tool needs to step in file discovery and replace the paths to the `*.ipynb` files with an intermediate python representation of the notebook, use `flake8` on the intermediate files and map the errors back to their origin in the notebook. The only way I found to achieve this is by using `Application` in a way it isn't intended to be used.

Since `Application` isn't part of the official API, the internals also aren't as stable as an API would be, which is why I have [scheduled tests against the default branch of `flake8`](https://github.com/s-weigand/flake8-nb/actions/workflows/test-nightly-schedule.yml) and supported `flake8` versions have an [upper bound to the tested versions](https://github.com/s-weigand/flake8-nb/blob/2a4c46085e73ad3c5b3a6603a68c6076ee7e3050/setup.py#L19).

If you have a closer look at how the properties of `Application` are overwritten (monkey patching of runtime objects), you might understand why I think that `hack` is an appropriate term.

That said, being a `hack` doesn't mean it is bad software since it works as intended, but it denotes that  `Application` is used in a way it isn't intended to be used.

**Comment by StefanBrand** [comment]
Ok, I understand better now, thanks for your explanation!

PS.: I've decided to go with `nbqa` as it supports both flake8 and black (and many other tools), which is quite comfortable for my use-case. :)

**Comment by MarcoGorelli** [comment]
Hi @StefanBrand - I'm glad you've found nbqa useful, though note that it's even more of a hack than flake8-nb :flushed:

**Comment by s-weigand** [comment]
@MarcoGorelli I wouldn't say nbQA is a hack, but a very sophisticated parsing wrapper around the tools which it supports. At least we don't monkey patch objects at runtime there 😝
